automata/pda/pda.py

```diff
@@ -24,6 +40,280 @@ class PDA(Automaton, metaclass=abc.ABCMeta):
+    ) -> Tuple[str, str, str]:
+        if (
+            from_state.remaining_input == to_state.remaining_input
+            or len(from_state.remaining_input) == 0
+        ):
+            input_symbol = ""
+        else:
+            input_symbol = from_state.remaining_input[0]
+
+        stack_top_symbol = from_state.stack.top()
+
+        if len(from_state.stack) == len(to_state.stack) + 1:
+            stack_push_symbols = ""
+        else:
+            stack_push_symbols = "".join(to_state.stack[len(from_state.stack) - 1 :])
+
+        return (input_symbol, stack_top_symbol, stack_push_symbols[::-1])
+
+    @abc.abstractmethod
+    def iter_transitions(
+        self,
+    ) -> Generator[Tuple[PDAStateT, PDAStateT, Tuple[str, str, str]], None, None]:
+        """
+        Iterate over all transitions in the automaton.
+        Each transition is a tuple of the form
+        (from_state, to_state, (input_symbol, stack_top_symbol, stack_push_symbols))
+        """
+
+    def show_diagram(
+        self,
+        input_str: Optional[str] = None,
+        path: Union[str, os.PathLike, None] = None,
+        *,
+        layout_method: LayoutMethod = "dot",
+        horizontal: bool = True,
+        reverse_orientation: bool = False,
+        fig_size: Union[Tuple[float, float], Tuple[float], None] = None,
+        font_size: float = 14.0,
+        arrow_size: float = 0.85,
+        state_separation: float = 0.5,
+    ) -> pgv.AGraph:
+        """
+        Generates the graph associated with the given PDA.
+        Args:
+            input_str (str, optional): String list of input symbols. Defaults to None.
+            - path (str or os.PathLike, optional): Path to output file. If
+              None, the output will not be saved.
+            - horizontal (bool, optional): Direction of node layout. Defaults
+              to True.
+            - reverse_orientation (bool, optional): Reverse direction of node
+              layout. Defaults to False.
+            - fig_size (tuple, optional): Figure size. Defaults to None.
+            - font_size (float, optional): Font size. Defaults to 14.0.
+            - arrow_size (float, optional): Arrow head size. Defaults to 0.85.
+            - state_separation (float, optional): Node distance. Defaults to 0.5.
+        Returns:
+            AGraph corresponding to the given automaton.
+        """
+
+        if not _visual_imports:
+            raise ImportError(
+                "Missing visualization packages; "
+                "please install coloraide and pygraphviz."
+            )
+
+        # Defining the graph.
+        graph = pgv.AGraph(strict=False, directed=True)
+
+        if fig_size is not None:
+            graph.graph_attr.update(size=", ".join(map(str, fig_size)))
+
+        graph.graph_attr.update(ranksep=str(state_separation))
+        font_size_str = str(font_size)
+        arrow_size_str = str(arrow_size)
+
+        if horizontal:
```

Comment:
Nit: Could this layout determination logic be put in a helper function? This seems like it's identical to what is being used in the FA class.